### ISSUE TITLE
Data Augmentation works !

### ISSUE THREAD

**Issue by abelsalm** [ivadomed/lumbar-classification-rsna-challenge-2024#38]
Since we wanted to try data augmentation for a long time and we knew the ResNet50 architecture we were using was quickly overfitting the data, I tried a simple experiment : 

I trained two different models on SCS classification using the patches extracted with a constant physical volume : 

- The first was the control sample. I splitted the data with 75% for training and 25% for validation (using a random seed i noted) and applied the data a series of basic transformation 
`            LoadImaged(keys=['image']),  # load
            Spacingd(keys=['image'], pixdim=(0.4, 0.4, 4.4), mode=('bilinear')),  # resample
            EnsureChannelFirstd(keys=["image"]), 
            SpatialPadd(keys=['image'], spatial_size=(120, 80, 6)),  # fixed padding (just to be sure it is big enough)
            CenterSpatialCropd(keys=['image'], roi_size=(120, 80, 6)),  # crop centered
            ScaleIntensityd(keys=['image']),  
            NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True),  
            ToTensord(keys=['image']) `

- The second one used data augmentation. I generated the same train/val split as the one on the first model using the random seed. But this time, the first series of transformation was applied as in the first model, so it created 9552 for the training as before, but also another series was applied with different operations, to double the amount of images 
`            LoadImaged(keys=['image']),  # load
            Spacingd(keys=['image'], pixdim=(0.4, 0.4, 4.4), mode=('bilinear')),  # resample
            EnsureChannelFirstd(keys=["image"]),  
            RandRotated(keys=['image'], prob=1, range_x=0.2), # add random rotation between -0.2 and 0.2 radians with probability 1
            SpatialPadd(keys=['image'], spatial_size=(120, 80, 6)),  # same padding
            RandSpatialCropd(keys=['image'], roi_size=(120, 80, 6), random_size=False),  # randomize the crop so that now it won't be centered 
            ScaleIntensityd(keys=['image']),  
            NormalizeIntensityd(keys=['image'], nonzero=True, channel_wise=True), 
            ToTensord(keys=['image']) `

This mean the second model was trained with the same first 9552 images and additional 9552 other images obtained from slightly different transforms, with twice the amount of data. 

Here is the result for the first model (sorry the plot is a bit messy) : 
![training_loss_scs_model_layers_ 3, 4, 6, 3 _epochs_30_lr_0 0001_augmentation_False_wd_0 0001](https://github.com/user-attachments/assets/d7240219-910f-41bb-a08c-49a3d7d9c16e)

Here is the result with data augmentation : 
![training_loss_scs_model_layers_ 3, 4, 6, 3 _epochs_30_lr_0 0001_augmentation_True_wd_0 0001](https://github.com/user-attachments/assets/910789d6-221f-41f3-ba2a-2c3037a57942)

Even if there are random processes involved in the training, we have exactly the same images in the validation set, so the only random process are the initialization of the weights and the decay...

We observe that the model doesn't overfit within 30 epochs compared to less than 10 for the control and that the best validation loss was improved from 0.35 to 0.26 (may be it could have been lower with moire epochs).

I find this very promising ! Teams in the top 10 had validation loss around 0.25/0.26 so w'ere getting close here.

**Comment by abelsalm** [comment]
Here is the same training training procediure as the second plot, but done for 60 epochs...
![training_loss_scs_model_layers_ 3, 4, 6, 3 _epochs_60_lr_0 0001_augmentation_True_wd_0 0001](https://github.com/user-attachments/assets/e28bc975-f9c6-4c49-b59e-12802e0ce6f7)

Any thoughts about overfitting ? @tomDag25 @jcohenadad 
It's just that without data augmentation overfitting appears quickly in almost 10 epochs, but here simply by doubling the amount of data we get this figure... So I'm doubting on the procedure I used...

To remind the procedure data has been generated this way : 

```#code used in training script for data processing
     seed = 42
     generator = torch.Generator().manual_seed(seed)
     data = prepare_data(data_dir, csv_file, transform1) # transform without random crop and rotate
     # this is the split for the first training without data augmentation
     train_dataset, val_dataset = torch.utils.data.random_split(data, [train_size, val_size], generator=generator)
    
     # this is the second part of the data
     data_aug = prepare_data(data_dir, csv_file, transform2) # where transform2 includes random crops and rotation
     train_dataset_aug, val_dataset_aug = torch.utils.data.random_split(data_aug, [train_size, val_size], generator=generator)
    
     # total data used for training
     training_set = ConcateDataset([train_dataset, train_dataset_aug]) # combine original and data augmentation
    
     # total data used for validation, we don't use val_dataset_aug
     # validation is done on the same set as the training without data augmentation
     val_set = val_dataset ```

**Comment by abelsalm** [comment]
**Update : it indeed improves the results, but not as much as I thought.** 

I was before creating the splits among the images as a whole shuffled dataset, not the subjects. It could be the reason why the model would fit so well on the validation set. This procedure was not correct.

So i created a new split, now among the subjects, splitting 1/4 of the subjects for validation and 3/4 for the training. It's on duke on 20250102nii_data_splits. 

Training with simple data augmentation on this split, I couldn't reach **less than 0.3**, which is already good, but could be still improved regarding the best team's results. 
![Image](https://github.com/user-attachments/assets/20dcc07c-a4c1-4942-81db-cdd4e763fbc0)

I think the next step would be to add other data augmentation transformations but the real game changer could be **MixUp** and **CutMix** techniques. I'm going to see if I can implement this with MONAI. 

@tomDag25
